token/token.go

```diff
@@ -447,15 +461,366 @@ func (mgm *tokenManager) GenerateServiceAccountToken(req *goa.RequestData, saID
+	claims["exp"] = kcClaims.ExpiresAt
+	claims["nbf"] = kcClaims.NotBefore
+	claims["iat"] = kcClaims.IssuedAt
+	claims["iss"] = kcClaims.Issuer
+	claims["aud"] = kcClaims.Audience
+	claims["typ"] = "Refresh"
+	claims["auth_time"] = 0
+
+	if identity != nil {
+		claims["sub"] = identity.ID.String()
+	} else {
+		claims["sub"] = kcClaims.Subject
+	}
+
+	claims["azp"] = kcClaims.Audience
+	claims["session_state"] = kcClaims.SessionState
+
+	return token, nil
+}
+
+// GenerateUnsignedUserRefreshTokenForIdentity generates an unsigned OAuth2 user refresh token for the given identity
+func (mgm *tokenManager) GenerateUnsignedUserRefreshTokenForIdentity(ctx context.Context, identity account.Identity) (*jwt.Token, error) {
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Header["kid"] = mgm.userAccountPrivateKey.KeyID
+
+	req := goa.ContextRequest(ctx)
+	if req == nil {
+		return nil, errors.New("missing request in context")
+	}
+
+	authOpenshiftIO := rest.AbsoluteURL(req, "", mgm.config)
+	openshiftIO, err := rest.ReplaceDomainPrefixInAbsoluteURL(req, "", "", mgm.config)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	claims := token.Claims.(jwt.MapClaims)
+	claims["jti"] = uuid.NewV4().String()
+	iat := time.Now().Unix()
+	exp := iat + 30*24*60*60 // In 30 days
+	claims["exp"] = exp
+	claims["nbf"] = 0
+	claims["iat"] = iat
+	claims["iss"] = authOpenshiftIO
+	claims["aud"] = openshiftIO
+	claims["typ"] = "Refresh"
+	claims["auth_time"] = 0
+	claims["sub"] = identity.ID.String()
+
+	return token, nil
+}
+
+// ConvertTokenSet converts the token set to oauth2.Token
+func (mgm *tokenManager) ConvertTokenSet(tokenSet TokenSet) *oauth2.Token {
+	var accessToken, refreshToken, tokenType string
+	extra := make(map[string]interface{})
+	if tokenSet.AccessToken != nil {
+		accessToken = *tokenSet.AccessToken
+	}
+	if tokenSet.RefreshToken != nil {
+		refreshToken = *tokenSet.RefreshToken
+	}
+	if tokenSet.TokenType != nil {
+		tokenType = *tokenSet.TokenType
+	}
+	var expire time.Time
+	if tokenSet.ExpiresIn != nil {
+		expire = time.Now().Add(time.Duration(*tokenSet.ExpiresIn) * time.Second)
+		extra["expires_in"] = *tokenSet.ExpiresIn
+	}
+	if tokenSet.RefreshExpiresIn != nil {
+		extra["refresh_expires_in"] = *tokenSet.RefreshExpiresIn
+	}
+	if tokenSet.NotBeforePolicy != nil {
+		extra["not_before_policy"] = *tokenSet.NotBeforePolicy
+	}
+
+	oauth2Token := &oauth2.Token{
+		AccessToken:  accessToken,
+		RefreshToken: refreshToken,
+		TokenType:    tokenType,
+		Expiry:       expire,
+	}
+	oauth2Token = oauth2Token.WithExtra(extra)
+
+	return oauth2Token
+}
+
+// ConvertToken converts the oauth2.Token to a token set
+func (mgm *tokenManager) ConvertToken(oauthToken oauth2.Token) (*TokenSet, error) {
+
+	tokenSet := &TokenSet{
+		AccessToken:  &oauthToken.AccessToken,
+		RefreshToken: &oauthToken.RefreshToken,
+		TokenType:    &oauthToken.TokenType,
+	}
+
+	var err error
+	tokenSet.ExpiresIn, err = mgm.extraInt(oauthToken, "expires_in")
```

Comment:
So, you want to move all claim names to conts? There are a lot of them :) I'm afraid then it would be harder to read the code. And we should be good in terms of possible typos because they are all covered by tests.